### ISSUE TITLE
Very slow performance using SeekToCurrentErrorHandler with manual_immediate ack-mode

### ISSUE THREAD

**Issue by rlanhellas** [spring-projects/spring-kafka#1522]
**Affects Version(s):**  Spring Kafka 2.5.2.RELEASE

I'm trying to create a consumer that can consume a multiple schema topic (on the same topic I have multiple Avro schemas). Everything works fine using the following configuration:

1- In my application.properties, I have following:
```
spring.kafka.consumer.value-deserializer=org.springframework.kafka.support.serializer.ErrorHandlerDeserializer
spring.kafka.consumer.properties.spring.deserializer.value.delegate.class=io.confluent.kafka.serializers.KafkaAvroDeserializer
spring.kafka.consumer.enable-auto-commit=false
spring.kafka.listener.ack-mode=manual_immediate
```

2- I have an ErrorHandler configuration to check if this error should be retried or just ignored (commit and go to next record):
```
//Inside configuration class ...
factory.setErrorHandler(new MyCustomErrorHandler());
```

3- MyCustomerErrorHandler class
```
public class MyCustomErrorHandler extends SeekToCurrentErrorHandler {
   ...
   @Override
   public void handle(...) {
      SeekUtils.seekOrRecover(exception, records, consumer, container, checkIfShouldCommit(exception), 
      this.getSkipPredicate(records,exception), this.logger, this.getLogLevel());
   }
}
```

Note that I have a method "checkIfShouldCommit" that returns true or false. If exception is derived from "DeserializationException" I  just return true and commit the message. All this works fine, but with very slow performance.

**Now, I have a big performance problem:**
I made a test with 10K  records, so I produced 10K messages that will throw DeserializationException, to test my logic. The consumer spent about 8 minutes to consume 10K messages, **very very slow.** 

So, searching for the problem I found 2 possible causes:

1. ack-mode manual_immediate is about 10x times slower than the ack-mode manual. I made a test with normal records (without Deserialization errors) and ack-mode manual processed about 10K records in less than 1 second, very fast. But using manual_immediate mode I got about 10 seconds. 
2. SeekToCurrentErrorHandler gets the current record, commit and discard remaining records, to get in the next poll again. Think about: If I have 10K records (all with erros) and max.poll.records is 500, I always get 500 records, commit the first and discard 499, to get 500 again in the next poll. 
3. SeekToCurrentErrorHandler don't let me use ack-mode manual, just manual_immediate. So, putting all together (item 2 and 3) I got a very very slow consumer.

**Comment by garyrussell** [comment]
The slow down is expected since you are doing commits for each record instead of for the batch (with MANUAL).

Have you tried setting the container property `syncCommits` to `false`? You can add a commit callback to get the results (or it will just be logged by default).

**Comment by rlanhellas** [comment]
Hi @garyrussell , thanks for your reply. Yes, I updated the syncCommits to false and tried again, but I didn't note any performance improvement. This 10K records, with errors, continue with about 8-10 minutes to process everything.

I understand that MANUAL with much more faster, but I can't use this with SeekToCurrentErrorHandler.

**Comment by rlanhellas** [comment]
Thanks @garyrussell , I will wait for next Spring Kafka release to use this feature. But is worth to note that I found another way to improve performance:

Changing MyCustomErrorHandler to implements directly ErrorHandler and just ignore the message will give the power to use ack mode MANUAL.

```
public class MyCustomErrorHandler implements ErrorHandler {
   ...
   @Override
   public void handle(...) {
      //Do nothing, just log if needed
   }
}
```

In this way Spring Kafka will get 10K records (with errors), just ignore then, without commit. When I got the first committed message after this 10K records, all behind will be committed. 

Of course, if I have 1 million messages with errors, without commit, and my application goes down, I will need to process again, but I think this tradeoff is ok because my performance is amazing when using ackmode  MANUAL. 

I suggest for people just use ack mode MANUAL_IMMEDIATE if the number of messages is small or performance is not mandatory.
